source/common/router/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,191 @@
+      validator_(factory_context.messageValidationContext().dynamicValidationVisitor()),
+      init_target_(fmt::format("FilterConfigSubscription init {}", filter_config_name_),
+                   [this]() { start(); }),
+      scope_(factory_context.scope().createScope(stat_prefix + "filter_config_discovery." +
+                                                 filter_config_name_ + ".")),
+      stat_prefix_(stat_prefix), filter_config_provider_manager_(filter_config_provider_manager),
+      subscription_id_(subscription_id) {
+  const auto resource_name = getResourceName();
+  subscription_ =
+      factory_context.clusterManager().subscriptionFactory().subscriptionFromConfigSource(
+          config_source, Grpc::Common::typeUrl(resource_name), *scope_, *this, resource_decoder_);
+}
+
+void FilterConfigSubscription::start() {
+  if (!started_) {
+    started_ = true;
+    subscription_->start({filter_config_name_});
+  }
+}
+
+void FilterConfigSubscription::onConfigUpdate(
+    const std::vector<Envoy::Config::DecodedResourceRef>& resources,
+    const std::string& version_info) {
+  // Make sure to make progress in case the control plane is temporarily inconsistent.
+  init_target_.ready();
+
+  if (resources.size() != 1) {
+    throw EnvoyException(fmt::format(
+        "Unexpected number of resources in FilterConfigDS response: {}", resources.size()));
+  }
+  const auto& filter_config = dynamic_cast<const envoy::config::core::v3::TypedExtensionConfig&>(
+      resources[0].get().resource());
+  if (filter_config.name() != filter_config_name_) {
+    throw EnvoyException(fmt::format("Unexpected resource name in FilterConfigDS response: {}",
+                                     filter_config.name()));
+  }
+  auto& factory = Envoy::Config::Utility::getAndCheckFactory<
+      Server::Configuration::NamedHttpFilterConfigFactory>(filter_config);
+  // Ensure that the factory is valid in the filter chain context.
+  for (auto* provider : filter_config_providers_) {
+    provider->validateConfig(factory);
+  }
+  ProtobufTypes::MessagePtr message = Envoy::Config::Utility::translateAnyToFactoryConfig(
+      filter_config.typed_config(), validator_, factory);
+  Http::FilterFactoryCb factory_callback =
+      factory.createFilterFactoryFromProto(*message, stat_prefix_, factory_context_);
+  ENVOY_LOG(debug, "Updating filter config {}", filter_config_name_);
+  for (auto* provider : filter_config_providers_) {
+    provider->onConfigUpdate(factory_callback, version_info);
+  }
+}
+
+void FilterConfigSubscription::onConfigUpdate(
+    const std::vector<Envoy::Config::DecodedResourceRef>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources, const std::string&) {
+  if (!removed_resources.empty()) {
+    ENVOY_LOG(error,
+              "Server sent a delta FilterConfigDS update attempting to remove a resource (name: "
+              "{}). Ignoring.",
+              removed_resources[0]);
+  }
```

Comment:
Did we discuss previously how delta xDS should look? I think what you have is consistent with RDS/EDS, but I'm also tempted to prefer to make delta xDS behave more consistently with new APIs. WDYT? CC @markdroth

Reply:
We haven't discussed. I think for now it's simple since it's just a single resource per subscription. Eventually we might want to combine many resources per xDS request.